## DSC-NUCLIO_E01_DataPreparation
### PROFESOR_AlbertoVacas 
### PROJECT_BMWPricing
### GRUPO_03


### IMPORTAMOS LIBRERÍAS Y LEEMOS EL FICHERO


In [130]:
# IMPORTAMOS LIBRERÍAS NECESARIAS PARA EL ANÁLISIS
# pandas: para manipulación de datos
# numpy: para operaciones numéricas
# matplotlib y seaborn: para visualizaciones
# sklearn: para preprocesamiento de datos

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import OneHotEncoder, MinMaxScaler

# Configuramos el estilo de las visualizaciones
plt.style.use('seaborn-v0_8')
sns.set_palette("husl")


In [131]:
# LEEMOS EL FICHERO DE BMW PRICING
# El dataset contiene información sobre vehículos BMW usados con sus características y precios

df_bmw = pd.read_csv("bmw_pricing.csv", sep=",")
df_bmw.head()


marca modelo        km  potencia fecha_registro tipo_gasolina   color  \
0   NaN    118  140411.0     100.0     2012-02-01        diesel   black   
1   BMW     M4   13929.0     317.0            NaN        petrol    grey   
2   BMW    320  183297.0     120.0     2012-04-01        diesel   white   
3   BMW    420  128035.0     135.0            NaN        diesel     red   
4   BMW    425   97097.0     160.0            NaN        diesel  silver   

    tipo_coche volante_regulable aire_acondicionado camara_trasera  \
0          NaN              True               True          False   
1  convertible              True               True          False   
2          NaN             False              False          False   
3  convertible              True               True          False   
4          NaN              True               True          False   

  asientos_traseros_plegables elevalunas_electrico bluetooth   gps  \
0                         NaN                 True       NaN  True   
1                         NaN                False      True  True   
2                         NaN                 True     False  True   
3                         NaN                 True      True  True   
4                       False                False      True  True   

  alerta_lim_velocidad   precio fecha_venta  
0                  NaN  11300.0  2018-01-01  
1                 True  69700.0  2018-02-01  
2                False  10200.0  2018-02-01  
3                  NaN  25100.0  2018-02-01  
4                 True  33400.0  2018-04-01

### PARTE 1: LIMPIEZA Y PREPARACIÓN DE DATOS


#### ANÁLISIS INICIAL DEL DATASET


In [132]:
# Obtenemos información general del dataset: tipos de datos, número de filas y columnas
# Esto nos ayuda a entender la estructura inicial de los datos

df_bmw.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4843 entries, 0 to 4842
Data columns (total 18 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   marca                        3873 non-null   object 
 1   modelo                       4840 non-null   object 
 2   km                           4841 non-null   float64
 3   potencia                     4842 non-null   float64
 4   fecha_registro               2420 non-null   object 
 5   tipo_gasolina                4838 non-null   object 
 6   color                        4398 non-null   object 
 7   tipo_coche                   3383 non-null   object 
 8   volante_regulable            4839 non-null   object 
 9   aire_acondicionado           4357 non-null   object 
 10  camara_trasera               4841 non-null   object 
 11  asientos_traseros_plegables  1452 non-null   object 
 12  elevalunas_electrico         4841 non-null   object 
 13  bluetooth         

In [133]:
# Verificamos la forma del dataset (número de filas y columnas)
print(f"Dimensiones del dataset: {df_bmw.shape}")


Dimensiones del dataset: (4843, 18)


#### PREGUNTA 1: ¿QUÉ VARIABLES TIENEN NULOS?


In [134]:
# Identificamos qué columnas tienen valores nulos y cuántos
# Esto es crucial para decidir cómo manejar los valores faltantes

df_bmw.isnull().sum()


marca                           970
modelo                            3
km                                2
potencia                          1
fecha_registro                 2423
tipo_gasolina                     5
color                           445
tipo_coche                     1460
volante_regulable                 4
aire_acondicionado              486
camara_trasera                    2
asientos_traseros_plegables    3391
elevalunas_electrico              2
bluetooth                       728
gps                               0
alerta_lim_velocidad            728
precio                            6
fecha_venta                       1
dtype: int64

In [135]:
# Calculamos el porcentaje de nulos por columna para tener una mejor perspectiva
# Un porcentaje alto de nulos puede indicar que una columna no es útil para el análisis

porcentaje_nulos = (df_bmw.isnull().sum() / len(df_bmw)) * 100
porcentaje_nulos = porcentaje_nulos.sort_values(ascending=False)
porcentaje_nulos


asientos_traseros_plegables    70.018584
fecha_registro                 50.030973
tipo_coche                     30.146603
marca                          20.028908
alerta_lim_velocidad           15.032005
bluetooth                      15.032005
aire_acondicionado             10.035102
color                           9.188520
precio                          0.123890
tipo_gasolina                   0.103242
volante_regulable               0.082593
modelo                          0.061945
elevalunas_electrico            0.041297
km                              0.041297
camara_trasera                  0.041297
potencia                        0.020648
fecha_venta                     0.020648
gps                             0.000000
dtype: float64

#### ELIMINACIÓN INICIAL DE COLUMNAS

**Decisión:** Antes de proceder con el análisis, debemos identificar columnas que no aportan valor o que tienen demasiados nulos


In [136]:
# Analizamos la columna 'marca'. Si todos los valores son BMW o nulos, no aporta información
# La mayoría de los registros deberían ser BMW, pero hay algunos nulos
print(df_bmw['marca'].value_counts(dropna=False))


marca
BMW    3873
NaN     970
Name: count, dtype: int64


In [137]:
# Analizamos 'asientos_traseros_plegables' que tiene más del 70% de nulos
# Esta columna tiene demasiados valores faltantes para ser útil
print(df_bmw['asientos_traseros_plegables'].value_counts(dropna=False))


asientos_traseros_plegables
NaN      3391
False    1150
True      302
Name: count, dtype: int64


In [138]:
# Creamos una copia del dataset para trabajar sin modificar el original
df_bmw_clean = df_bmw.copy()

# DECISIÓN: Eliminamos columnas que no aportan valor:
# 1. 'marca': Todos los coches son BMW (o deberían serlo), no aporta información discriminativa
# 2. 'asientos_traseros_plegables': Tiene más del 70% de nulos, no es útil para el modelo

columnas_eliminar = ['marca', 'asientos_traseros_plegables']
df_bmw_clean = df_bmw_clean.drop(columns=columnas_eliminar)
df_bmw_clean.shape


(4843, 16)

#### CONVERSIÓN DE FECHAS A FORMATO DATETIME


In [139]:
# Convertimos las columnas de fecha a formato datetime para poder trabajar con ellas

df_bmw_clean['fecha_registro'] = pd.to_datetime(df_bmw_clean['fecha_registro'], errors='coerce')
df_bmw_clean['fecha_venta'] = pd.to_datetime(df_bmw_clean['fecha_venta'], errors='coerce')

# Verificamos la conversión
df_bmw_clean[['fecha_registro', 'fecha_venta']].info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4843 entries, 0 to 4842
Data columns (total 2 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   fecha_registro  2420 non-null   datetime64[ns]
 1   fecha_venta     4842 non-null   datetime64[ns]
dtypes: datetime64[ns](2)
memory usage: 75.8 KB


#### ELIMINACIÓN DE DUPLICADOS


In [140]:
# Identificamos registros duplicados
# Los duplicados pueden sesgar nuestro análisis y modelo

duplicados = df_bmw_clean.duplicated()
duplicados.sum()


np.int64(0)

In [141]:
# En el caso de tener duplicados los eliminaríamos y resetearíamos el índice
df_bmw_clean = df_bmw_clean.drop_duplicates().reset_index(drop=True)

# Luego podríamos calcular el número de registros eliminados
df_bmw.shape[0] - df_bmw_clean.shape[0]


0

#### CREACIÓN DE VARIABLES DERIVADAS


In [142]:
# Creamos una variable que puede ser útil para el modelo:
# 1. EDAD_VEHICULO: años desde el registro hasta la venta (puede ayudar a predecir precio)

# Calculamos la edad del vehículo en años al momento de la venta
df_bmw_clean['EDAD_VEHICULO'] = (df_bmw_clean['fecha_venta'] - df_bmw_clean['fecha_registro']).dt.days / 365

# Verificamos la nueva variable
df_bmw_clean['EDAD_VEHICULO'].describe()


count    2420.000000
mean        5.389322
std         2.528404
min        -5.504110
25%         4.079452
50%         4.838356
75%         5.835616
max        28.104110
Name: EDAD_VEHICULO, dtype: float64

#### PREGUNTA 2: MANEJO DE NULOS POR COLUMNA

Analizamos cada columna con nulos y decidimos la estrategia de imputación o eliminación.


In [143]:
# Revisamos nuevamente los nulos después de las transformaciones iniciales

nulos_actualizados = df_bmw_clean.isnull().sum()
nulos_actualizados = nulos_actualizados[nulos_actualizados > 0].sort_values(ascending=False)
nulos_actualizados


fecha_registro          2423
EDAD_VEHICULO           2423
tipo_coche              1460
bluetooth                728
alerta_lim_velocidad     728
aire_acondicionado       486
color                    445
precio                     6
tipo_gasolina              5
volante_regulable          4
modelo                     3
km                         2
elevalunas_electrico       2
camara_trasera             2
potencia                   1
fecha_venta                1
dtype: int64

**Estrategia para nulos por columna:**

1. **precio (target)**: Eliminaremos registros con precio nulo ya que no podemos predecir sin target
2. **modelo/km/potencia**: Son variables numéricas críticas. Tienen pocos nulos. Tiene sentido eliminar o imputar con mediana/moda según corresponda
3. **fecha_registro**: Tiene aprox 50% nulos. Tiene sentido crear variable binaria "fecha_registro_disponible" o imputar con mediana
4. **tipo_gasolina**: Imputaremos con moda (valor más frecuente)


5. **tipo_coche**: Tiene aprox 30% nulos. Tiene sentido imputar con moda o crear categoría "desconocido"
6. **color**: Tiene aprox 9% nulos. Tiene sentido imputar con moda o crear categoría "desconocido"
7. **bluetooth/alerta_lim_velocidad**: Tienen aprox 15% nulos. Tiene sentido tratar como False (como si fuera ausencia de característica)
8. **aire_acondicionado**: Tiene aprox 10% nulos. Tiene sentido imputar con moda


In [144]:
# 1. precio (target): Eliminamos registros sin precio
# No podemos usar registros sin precio para entrenar el modelo

print(f"Registros antes de eliminar nulos en precio: {len(df_bmw_clean)}")
df_bmw_clean = df_bmw_clean.dropna(subset=['precio'])
print(f"Registros despues: {len(df_bmw_clean)}")
print(f"Registros eliminados: {df_bmw.shape[0] - len(df_bmw_clean)}")


Registros antes de eliminar nulos en precio: 4843
Registros despues: 4837
Registros eliminados: 6


In [145]:
# 2. modelo, km, potencia
# Estas son variables importantes, eliminamos si faltan valores críticos

print("Nulos en variables numéricas críticas:")
print(f"modelo: {df_bmw_clean['modelo'].isnull().sum()}")
print(f"km: {df_bmw_clean['km'].isnull().sum()}")
print(f"potencia: {df_bmw_clean['potencia'].isnull().sum()}")

# Eliminamos registros sin modelo, km o potencia ya que són pocas
df_bmw_clean = df_bmw_clean.dropna(subset=['modelo', 'km', 'potencia'])
print(f"Registros después de eliminar nulos en variables críticas: {len(df_bmw_clean)}")


Nulos en variables numéricas críticas:
modelo: 3
km: 2
potencia: 1
Registros después de eliminar nulos en variables críticas: 4831


In [150]:
# 3. fecha_registro: Crear variable binaria y mantener nulos
# Como tiene muchos nulos (aprox 50%), creamos una variable indicadora
# Para EDAD_VEHICULO, los nulos se mantienen (serán NaN) y los trataremos después

df_bmw_clean['fecha_registro_disponible'] = df_bmw_clean['fecha_registro'].notna().astype(int)
df_bmw_clean['fecha_registro_disponible'].value_counts()

fecha_registro_disponible
1    2416
0    2415
Name: count, dtype: int64

In [153]:
# 4. tipo_gasolina: Imputamos con moda (valor más frecuente)
# Es una variable categórica importante

moda_tipo_gasolina = df_bmw_clean['tipo_gasolina'].mode()[0]
print(f"Moda de tipo_gasolina: {moda_tipo_gasolina}")
print(f"Nulos antes: {df_bmw_clean['tipo_gasolina'].isnull().sum()}")

df_bmw_clean['tipo_gasolina'] = df_bmw_clean['tipo_gasolina'].fillna(moda_tipo_gasolina)
print(f"Nulos despues: {df_bmw_clean['tipo_gasolina'].isnull().sum()}")


Moda de tipo_gasolina: diesel
Nulos antes: 0
Nulos despues: 0


In [ ]:
# 5. MANEJO DE tipo_coche: Imputar con moda o crear categoría "desconocido"
# Tiene ~30% de nulos, pero es una variable importante

print("Valores únicos en tipo_coche:")
print(df_bmw_clean['tipo_coche'].value_counts(dropna=False))

# Decidimos crear categoría "desconocido" en lugar de imputar con moda
# porque los nulos pueden ser significativos (coches sin tipo específico)
df_bmw_clean['tipo_coche'] = df_bmw_clean['tipo_coche'].fillna('desconocido')
print(f"\nNulos después: {df_bmw_clean['tipo_coche'].isnull().sum()}")


In [ ]:
# 6. MANEJO DE color: Imputar con moda o crear categoría "desconocido"
# Tiene ~9% de nulos

print("Valores únicos en color (top 10):")
print(df_bmw_clean['color'].value_counts(dropna=False).head(10))

# Creamos categoría "desconocido" para mantener la información de que falta el dato
df_bmw_clean['color'] = df_bmw_clean['color'].fillna('desconocido')
print(f"\nNulos después: {df_bmw_clean['color'].isnull().sum()}")


In [ ]:
# 7. MANEJO DE VARIABLES BOOLEANAS: bluetooth, alerta_lim_velocidad, aire_acondicionado
# Estas variables tienen valores True/False/NaN. Tratamos NaN como False (ausencia de característica)

variables_booleanas = ['bluetooth', 'alerta_lim_velocidad', 'aire_acondicionado', 
                       'volante_regulable', 'camara_trasera', 'elevalunas_electrico']

for var in variables_booleanas:
    nulos_antes = df_bmw_clean[var].isnull().sum()
    # Convertimos a booleano: True/False se mantienen, NaN se convierte en False
    df_bmw_clean[var] = df_bmw_clean[var].fillna(False)
    # Aseguramos que sean booleanos
    df_bmw_clean[var] = df_bmw_clean[var].astype(bool)
    print(f"{var}: {nulos_antes} nulos imputados como False")


In [ ]:
# 8. MANEJO DE gps: Ya es booleano, verificamos que no tenga nulos
print(f"gps - nulos: {df_bmw_clean['gps'].isnull().sum()}")
print(f"gps - tipo: {df_bmw_clean['gps'].dtype}")


In [ ]:
# 9. MANEJO DE fecha_venta: Eliminar si falta (es crítica)
df_bmw_clean = df_bmw_clean.dropna(subset=['fecha_venta'])
print(f"Registros después de eliminar nulos en fecha_venta: {len(df_bmw_clean)}")


In [ ]:
# Verificamos el estado final de los nulos
print("Estado final de nulos por columna:")
nulos_finales = df_bmw_clean.isnull().sum()
nulos_finales = nulos_finales[nulos_finales > 0]
if len(nulos_finales) > 0:
    print(nulos_finales)
else:
    print("¡No hay nulos restantes!")


#### ANÁLISIS Y LIMPIEZA DE OUTLIERS


In [ ]:
# Analizamos outliers en la variable target (precio)
# Los outliers pueden afectar significativamente el modelo

print("Estadísticas descriptivas de PRECIO:")
print(df_bmw_clean['precio'].describe())

# Visualizamos la distribución del precio
plt.figure(figsize=(12, 5))
plt.subplot(1, 2, 1)
df_bmw_clean['precio'].hist(bins=50)
plt.title('Distribución de Precio')
plt.xlabel('Precio (€)')
plt.ylabel('Frecuencia')

plt.subplot(1, 2, 2)
sns.boxplot(y=df_bmw_clean['precio'])
plt.title('Boxplot de Precio')
plt.ylabel('Precio (€)')

plt.tight_layout()
plt.show()


In [ ]:
# Identificamos y tratamos outliers en precio usando el método IQR
# Eliminamos precios <= 0 (no tienen sentido) y valores extremos

# Eliminamos precios <= 0
print(f"Registros con precio <= 0: {(df_bmw_clean['precio'] <= 0).sum()}")
df_bmw_clean = df_bmw_clean[df_bmw_clean['precio'] > 0]

# Método IQR para outliers superiores
Q1 = df_bmw_clean['precio'].quantile(0.25)
Q3 = df_bmw_clean['precio'].quantile(0.75)
IQR = Q3 - Q1

limite_inferior = Q1 - 1.5 * IQR
limite_superior = Q3 + 1.5 * IQR

print(f"\nLímites IQR:")
print(f"Q1: {Q1:.2f}, Q3: {Q3:.2f}, IQR: {IQR:.2f}")
print(f"Límite inferior: {limite_inferior:.2f}")
print(f"Límite superior: {limite_superior:.2f}")

outliers_inferiores = (df_bmw_clean['precio'] < limite_inferior).sum()
outliers_superiores = (df_bmw_clean['precio'] > limite_superior).sum()

print(f"\nOutliers inferiores: {outliers_inferiores}")
print(f"Outliers superiores: {outliers_superiores}")

# DECISIÓN: Para precios, los outliers superiores pueden ser válidos (coches de lujo)
# Solo eliminamos outliers inferiores (precios anormalmente bajos)
df_bmw_clean = df_bmw_clean[df_bmw_clean['precio'] >= limite_inferior]
print(f"\nRegistros después de eliminar outliers inferiores: {len(df_bmw_clean)}")


In [ ]:
# Analizamos outliers en otras variables numéricas: km, potencia, EDAD_VEHICULO

variables_numericas = ['km', 'potencia', 'EDAD_VEHICULO']

for var in variables_numericas:
    print(f"\n=== Análisis de {var} ===")
    print(df_bmw_clean[var].describe())
    
    # Eliminamos valores negativos si los hay (no tienen sentido)
    if (df_bmw_clean[var] < 0).any():
        print(f"Valores negativos encontrados: {(df_bmw_clean[var] < 0).sum()}")
        df_bmw_clean = df_bmw_clean[df_bmw_clean[var] >= 0]
    
    # Para km y potencia, eliminamos valores extremos superiores (pueden ser errores)
    if var in ['km', 'potencia']:
        Q1 = df_bmw_clean[var].quantile(0.25)
        Q3 = df_bmw_clean[var].quantile(0.75)
        IQR = Q3 - Q1
        limite_superior = Q3 + 3 * IQR  # Usamos 3*IQR para ser menos restrictivos
        
        outliers = (df_bmw_clean[var] > limite_superior).sum()
        print(f"Outliers superiores (> Q3 + 3*IQR): {outliers}")
        
        # Solo eliminamos si hay muchos outliers (más del 1%)
        if outliers > len(df_bmw_clean) * 0.01:
            df_bmw_clean = df_bmw_clean[df_bmw_clean[var] <= limite_superior]
            print(f"Eliminados {outliers} outliers")
    
    print(f"Registros restantes: {len(df_bmw_clean)}")


#### GUARDAMOS EL DATASET LIMPIO


In [ ]:
# Guardamos el dataset limpio en formato pickle para uso posterior
df_bmw_clean.to_pickle("df_bmw_clean.pkl")

print(f"Dataset guardado. Dimensiones finales: {df_bmw_clean.shape}")
print(f"Registros originales: {df_bmw.shape[0]}")
print(f"Registros finales: {df_bmw_clean.shape[0]}")
print(f"Registros eliminados: {df_bmw.shape[0] - df_bmw_clean.shape[0]}")


### PARTE 2: ANÁLISIS EXPLORATORIO DE DATOS (EDA)


#### LEEMOS EL DATASET LIMPIO


In [ ]:
# Leemos el dataset limpio
df_bmw_clean = pd.read_pickle("df_bmw_clean.pkl")
df_bmw_clean.head()


#### PREGUNTA 3: ANÁLISIS UNIVARIABLE

Analizamos cada variable individualmente para entender su distribución y encontrar información interesante.


In [ ]:
# ANÁLISIS DE VARIABLES NUMÉRICAS

variables_numericas = ['km', 'potencia', 'precio', 'EDAD_VEHICULO']

for var in variables_numericas:
    print(f"\n{'='*50}")
    print(f"ANÁLISIS DE {var.upper()}")
    print(f"{'='*50}")
    print(df_bmw_clean[var].describe())
    
    # Visualizaciones
    fig, axes = plt.subplots(1, 2, figsize=(14, 5))
    
    # Histograma
    axes[0].hist(df_bmw_clean[var].dropna(), bins=50, edgecolor='black')
    axes[0].set_title(f'Distribución de {var}')
    axes[0].set_xlabel(var)
    axes[0].set_ylabel('Frecuencia')
    
    # Boxplot
    axes[1].boxplot(df_bmw_clean[var].dropna())
    axes[1].set_title(f'Boxplot de {var}')
    axes[1].set_ylabel(var)
    
    plt.tight_layout()
    plt.show()
    
    # Información interesante
    if var == 'precio':
        print(f"\n💡 INSIGHT: El precio medio es {df_bmw_clean[var].mean():.2f}€")
        print(f"   El precio mediano es {df_bmw_clean[var].median():.2f}€")
        print(f"   La diferencia entre media y mediana sugiere una distribución sesgada")
    elif var == 'km':
        print(f"\n💡 INSIGHT: Los coches tienen en promedio {df_bmw_clean[var].mean():.0f} km")
        print(f"   El rango va desde {df_bmw_clean[var].min():.0f} hasta {df_bmw_clean[var].max():.0f} km")
    elif var == 'potencia':
        print(f"\n💡 INSIGHT: La potencia media es {df_bmw_clean[var].mean():.1f} CV")
        print(f"   Potencia mínima: {df_bmw_clean[var].min():.0f} CV, máxima: {df_bmw_clean[var].max():.0f} CV")


In [ ]:
# ANÁLISIS DE VARIABLES CATEGÓRICAS

variables_categoricas = ['modelo', 'tipo_gasolina', 'color', 'tipo_coche']

for var in variables_categoricas:
    print(f"\n{'='*50}")
    print(f"ANÁLISIS DE {var.upper()}")
    print(f"{'='*50}")
    
    value_counts = df_bmw_clean[var].value_counts()
    print(f"\nTop 10 valores más frecuentes:")
    print(value_counts.head(10))
    
    print(f"\nNúmero de categorías únicas: {df_bmw_clean[var].nunique()}")
    print(f"Porcentaje del valor más frecuente: {(value_counts.iloc[0] / len(df_bmw_clean)) * 100:.2f}%")
    
    # Visualización
    plt.figure(figsize=(12, 6))
    if df_bmw_clean[var].nunique() > 20:
        # Si hay muchas categorías, mostramos solo las top 15
        top_values = value_counts.head(15)
        top_values.plot(kind='bar')
        plt.title(f'Top 15 valores más frecuentes en {var}')
    else:
        value_counts.plot(kind='bar')
        plt.title(f'Distribución de {var}')
    
    plt.xlabel(var)
    plt.ylabel('Frecuencia')
    plt.xticks(rotation=45, ha='right')
    plt.tight_layout()
    plt.show()
    
    # Insights
    if var == 'modelo':
        print(f"\n💡 INSIGHT: El modelo más común es '{value_counts.index[0]}' con {value_counts.iloc[0]} unidades")
    elif var == 'tipo_gasolina':
        print(f"\n💡 INSIGHT: {value_counts.index[0]} es el tipo de combustible más común")
        print(f"   Distribución: {dict(value_counts)}")
    elif var == 'color':
        print(f"\n💡 INSIGHT: El color más popular es '{value_counts.index[0]}'")


In [ ]:
# ANÁLISIS DE VARIABLES BOOLEANAS

variables_booleanas = ['volante_regulable', 'aire_acondicionado', 'camara_trasera', 
                       'elevalunas_electrico', 'bluetooth', 'gps', 'alerta_lim_velocidad']

print("Distribución de variables booleanas:")
print("="*60)

for var in variables_booleanas:
    value_counts = df_bmw_clean[var].value_counts()
    porcentaje_true = (value_counts.get(True, 0) / len(df_bmw_clean)) * 100
    print(f"\n{var}:")
    print(f"  True: {value_counts.get(True, 0)} ({porcentaje_true:.1f}%)")
    print(f"  False: {value_counts.get(False, 0)} ({100-porcentaje_true:.1f}%)")

# Visualización
fig, axes = plt.subplots(2, 4, figsize=(16, 8))
axes = axes.flatten()

for i, var in enumerate(variables_booleanas):
    value_counts = df_bmw_clean[var].value_counts()
    axes[i].bar(value_counts.index.astype(str), value_counts.values)
    axes[i].set_title(var)
    axes[i].set_ylabel('Frecuencia')

plt.tight_layout()
plt.show()


#### PREGUNTA 4: ANÁLISIS DE CORRELACIÓN INICIAL

Identificamos variables correlacionadas que podrían causar multicolinealidad en el modelo.


In [ ]:
# Calculamos la matriz de correlación solo para variables numéricas
# La correlación mide la relación lineal entre variables

variables_numericas_corr = ['km', 'potencia', 'precio', 'EDAD_VEHICULO', 'fecha_registro_disponible']
corr = df_bmw_clean[variables_numericas_corr].corr()

print("Matriz de correlación (variables numéricas):")
print(corr)

# Visualización con mapa de calor
plt.figure(figsize=(10, 8))
sns.heatmap(corr, annot=True, cmap='coolwarm', center=0, 
            square=True, linewidths=1, cbar_kws={"shrink": 0.8}, fmt='.2f')
plt.title('Matriz de Correlación - Variables Numéricas')
plt.tight_layout()
plt.show()


In [ ]:
# Identificamos pares de variables con alta correlación (|r| > 0.7)
# Estas variables pueden causar multicolinealidad

print("Pares de variables con correlación alta (|r| > 0.7):")
print("="*60)

# Obtenemos el triángulo superior de la matriz (sin la diagonal)
mask = np.triu(np.ones_like(corr, dtype=bool), k=1)
corr_triangulo = corr.where(mask)

# Buscamos correlaciones altas
for i in range(len(corr_triangulo.columns)):
    for j in range(i+1, len(corr_triangulo.columns)):
        valor = corr_triangulo.iloc[i, j]
        if not np.isnan(valor) and abs(valor) > 0.7:
            var1 = corr_triangulo.columns[i]
            var2 = corr_triangulo.columns[j]
            print(f"{var1} - {var2}: {valor:.3f}")

# Si no hay correlaciones altas
if corr_triangulo.abs().max().max() <= 0.7:
    print("No se encontraron correlaciones altas (|r| > 0.7) entre variables numéricas")
    print(f"Correlación máxima: {corr_triangulo.abs().max().max():.3f}")


#### PREGUNTA 5: ANÁLISIS VARIABLE VS TARGET

Analizamos la relación entre cada variable y el precio (target) para identificar insights interesantes.


In [ ]:
# ANÁLISIS: Variables numéricas vs Precio

variables_numericas_vs_target = ['km', 'potencia', 'EDAD_VEHICULO']

for var in variables_numericas_vs_target:
    print(f"\n{'='*50}")
    print(f"ANÁLISIS: {var.upper()} vs PRECIO")
    print(f"{'='*50}")
    
    # Correlación
    correlacion = df_bmw_clean[var].corr(df_bmw_clean['precio'])
    print(f"Correlación con precio: {correlacion:.3f}")
    
    # Visualización
    fig, axes = plt.subplots(1, 2, figsize=(14, 5))
    
    # Scatter plot
    axes[0].scatter(df_bmw_clean[var], df_bmw_clean['precio'], alpha=0.5)
    axes[0].set_xlabel(var)
    axes[0].set_ylabel('Precio (€)')
    axes[0].set_title(f'{var} vs Precio')
    
    # Boxplot por cuartiles
    df_bmw_clean[f'{var}_quartile'] = pd.qcut(df_bmw_clean[var], q=4, labels=['Q1', 'Q2', 'Q3', 'Q4'], duplicates='drop')
    sns.boxplot(data=df_bmw_clean, x=f'{var}_quartile', y='precio', ax=axes[1])
    axes[1].set_title(f'Precio por cuartiles de {var}')
    axes[1].set_xlabel(f'{var} (cuartiles)')
    axes[1].set_ylabel('Precio (€)')
    
    plt.tight_layout()
    plt.show()
    
    # Eliminamos la columna temporal
    df_bmw_clean.drop(columns=[f'{var}_quartile'], inplace=True)
    
    # Insight
    if abs(correlacion) > 0.3:
        direccion = "positiva" if correlacion > 0 else "negativa"
        print(f"\n💡 INSIGHT: Correlación {direccion} moderada/alta ({correlacion:.3f})")
        if var == 'km':
            print("   A mayor kilometraje, menor precio (correlación negativa esperada)")
        elif var == 'potencia':
            print("   A mayor potencia, mayor precio (correlación positiva esperada)")
    else:
        print(f"\n💡 INSIGHT: Correlación débil con precio ({correlacion:.3f})")


In [ ]:
# ANÁLISIS: Variables categóricas vs Precio

variables_categoricas_vs_target = ['modelo', 'tipo_gasolina', 'color', 'tipo_coche']

for var in variables_categoricas_vs_target:
    print(f"\n{'='*50}")
    print(f"ANÁLISIS: {var.upper()} vs PRECIO")
    print(f"{'='*50}")
    
    # Estadísticas por categoría
    stats_por_categoria = df_bmw_clean.groupby(var)['precio'].agg(['mean', 'median', 'count']).sort_values('mean', ascending=False)
    print(f"\nPrecio medio por categoría (top 10):")
    print(stats_por_categoria.head(10))
    
    # Visualización
    fig, axes = plt.subplots(1, 2, figsize=(16, 6))
    
    # Violin plot (si hay pocas categorías) o boxplot
    if df_bmw_clean[var].nunique() <= 15:
        # Violin plot para ver distribución
        top_categorias = stats_por_categoria.head(10).index
        df_top = df_bmw_clean[df_bmw_clean[var].isin(top_categorias)]
        sns.violinplot(data=df_top, x=var, y='precio', ax=axes[0])
        axes[0].set_title(f'Distribución de Precio por {var} (Top 10)')
        axes[0].set_xticklabels(axes[0].get_xticklabels(), rotation=45, ha='right')
    else:
        # Boxplot para muchas categorías
        top_categorias = stats_por_categoria.head(10).index
        df_top = df_bmw_clean[df_bmw_clean[var].isin(top_categorias)]
        sns.boxplot(data=df_top, x=var, y='precio', ax=axes[0])
        axes[0].set_title(f'Precio por {var} (Top 10)')
        axes[0].set_xticklabels(axes[0].get_xticklabels(), rotation=45, ha='right')
    
    # Precio medio por categoría
    top_10_mean = stats_por_categoria.head(10)
    top_10_mean['mean'].plot(kind='bar', ax=axes[1])
    axes[1].set_title(f'Precio Medio por {var} (Top 10)')
    axes[1].set_xlabel(var)
    axes[1].set_ylabel('Precio Medio (€)')
    axes[1].tick_params(axis='x', rotation=45)
    
    plt.tight_layout()
    plt.show()
    
    # Insight
    categoria_mas_cara = stats_por_categoria.index[0]
    categoria_mas_barata = stats_por_categoria.index[-1]
    diferencia = stats_por_categoria.loc[categoria_mas_cara, 'mean'] - stats_por_categoria.loc[categoria_mas_barata, 'mean']
    
    print(f"\n💡 INSIGHT:")
    print(f"   Categoría más cara: {categoria_mas_cara} (precio medio: {stats_por_categoria.loc[categoria_mas_cara, 'mean']:.2f}€)")
    print(f"   Categoría más barata: {categoria_mas_barata} (precio medio: {stats_por_categoria.loc[categoria_mas_barata, 'mean']:.2f}€)")
    print(f"   Diferencia: {diferencia:.2f}€")


In [ ]:
# ANÁLISIS: Variables booleanas vs Precio

variables_booleanas_vs_target = ['volante_regulable', 'aire_acondicionado', 'camara_trasera', 
                                  'elevalunas_electrico', 'bluetooth', 'gps', 'alerta_lim_velocidad']

print("Impacto de características booleanas en el precio:")
print("="*60)

for var in variables_booleanas_vs_target:
    stats = df_bmw_clean.groupby(var)['precio'].agg(['mean', 'median', 'count'])
    diferencia = stats.loc[True, 'mean'] - stats.loc[False, 'mean']
    porcentaje_diferencia = (diferencia / stats.loc[False, 'mean']) * 100
    
    print(f"\n{var}:")
    print(f"  Con {var}: {stats.loc[True, 'mean']:.2f}€ (n={stats.loc[True, 'count']})")
    print(f"  Sin {var}: {stats.loc[False, 'mean']:.2f}€ (n={stats.loc[False, 'count']})")
    print(f"  Diferencia: {diferencia:+.2f}€ ({porcentaje_diferencia:+.1f}%)")

# Visualización
fig, axes = plt.subplots(2, 4, figsize=(16, 8))
axes = axes.flatten()

for i, var in enumerate(variables_booleanas_vs_target):
    sns.boxplot(data=df_bmw_clean, x=var, y='precio', ax=axes[i])
    axes[i].set_title(var)
    axes[i].set_xlabel('')
    axes[i].set_ylabel('Precio (€)')

plt.tight_layout()
plt.show()


#### PREGUNTA 6: TRANSFORMACIÓN DE VARIABLES

Identificamos qué variables categóricas necesitan transformación y qué técnica usar.


In [ ]:
# Identificamos el tipo de cada variable para decidir la transformación adecuada

target = ['precio']  # Variable objetivo

def obtener_lista_variables(dataset, target):
    """
    Clasifica las variables en numéricas, booleanas y categóricas.
    """
    lista_numericas = []
    lista_boolean = []
    lista_categoricas = []
    
    for col in dataset.columns:
        if col in target:
            continue
        
        # Variables numéricas (float o int) con más de 2 valores únicos
        if dataset[col].dtype.kind in ('f', 'i') and len(dataset[col].unique()) > 2:
            lista_numericas.append(col)
        # Variables booleanas (bool o con solo 2 valores únicos)
        elif dataset[col].dtype == bool or (dataset[col].dtype.kind in ('f', 'i', 'O') and len(dataset[col].unique()) == 2):
            lista_boolean.append(col)
        # Variables categóricas (object o con pocos valores únicos)
        elif dataset[col].dtype.kind == 'O':
            lista_categoricas.append(col)
        # Variables datetime no se incluyen directamente
    
    return lista_numericas, lista_boolean, lista_categoricas

lista_numericas, lista_boolean, lista_categoricas = obtener_lista_variables(df_bmw_clean, target)

print("Clasificación de variables:")
print("="*60)
print(f"\nVariables NUMÉRICAS ({len(lista_numericas)}):")
for var in lista_numericas:
    print(f"  - {var}")

print(f"\nVariables BOOLEANAS ({len(lista_boolean)}):")
for var in lista_boolean:
    print(f"  - {var}")

print(f"\nVariables CATEGÓRICAS ({len(lista_categoricas)}):")
for var in lista_categoricas:
    n_categorias = df_bmw_clean[var].nunique()
    print(f"  - {var} ({n_categorias} categorías)")


In [ ]:
# Analizamos las variables categóricas para decidir la técnica de encoding

print("Análisis de variables categóricas para encoding:")
print("="*60)

for var in lista_categoricas:
    n_categorias = df_bmw_clean[var].nunique()
    distribucion = df_bmw_clean[var].value_counts()
    
    print(f"\n{var}:")
    print(f"  Número de categorías: {n_categorias}")
    print(f"  Top 5 categorías:")
    for cat, count in distribucion.head(5).items():
        porcentaje = (count / len(df_bmw_clean)) * 100
        print(f"    - {cat}: {count} ({porcentaje:.1f}%)")
    
    # Decisión de técnica
    if n_categorias <= 10:
        print(f"  → TÉCNICA: One-Hot Encoding (OHE) - pocas categorías")
    else:
        print(f"  → TÉCNICA: One-Hot Encoding (OHE) o Target Encoding - muchas categorías")
        print(f"    Considerar agrupar categorías poco frecuentes si hay muchas")


In [ ]:
# Aplicamos One-Hot Encoding a las variables categóricas
# OHE crea una columna binaria para cada categoría

print("Aplicando One-Hot Encoding a variables categóricas...")
print(f"Variables a transformar: {lista_categoricas}")

df_bmw_prep = pd.get_dummies(data=df_bmw_clean, columns=lista_categoricas, dtype=int)

print(f"\nDimensiones antes de OHE: {df_bmw_clean.shape}")
print(f"Dimensiones después de OHE: {df_bmw_prep.shape}")
print(f"Columnas nuevas creadas: {df_bmw_prep.shape[1] - df_bmw_clean.shape[1]}")

# Mostramos algunas columnas nuevas
print(f"\nEjemplo de columnas nuevas (primeras 10):")
nuevas_columnas = [col for col in df_bmw_prep.columns if col not in df_bmw_clean.columns]
print(nuevas_columnas[:10])


In [ ]:
# Normalizamos las variables numéricas usando MinMaxScaler
# MinMaxScaler escala los valores al rango [0, 1]
# Esto es importante para algoritmos que son sensibles a la escala

print("Normalizando variables numéricas con MinMaxScaler...")
print(f"Variables a normalizar: {lista_numericas}")

# Creamos el scaler
scaler = MinMaxScaler()

# Aplicamos la normalización
df_bmw_prep[lista_numericas] = scaler.fit_transform(df_bmw_prep[lista_numericas])

print("\nVerificación de la normalización:")
print(df_bmw_prep[lista_numericas].describe())

# Verificamos que los valores estén en el rango [0, 1]
print(f"\nRango de valores después de normalización:")
for var in lista_numericas:
    min_val = df_bmw_prep[var].min()
    max_val = df_bmw_prep[var].max()
    print(f"  {var}: [{min_val:.3f}, {max_val:.3f}]")


In [ ]:
# Verificamos el estado final del dataset preparado
print("Estado final del dataset preparado:")
print("="*60)
print(f"Dimensiones: {df_bmw_prep.shape}")
print(f"\nTipos de datos:")
print(df_bmw_prep.dtypes.value_counts())

print(f"\nPrimeras filas:")
df_bmw_prep.head()


#### GUARDAMOS EL DATASET PREPROCESADO


In [ ]:
# Guardamos el dataset preprocesado
df_bmw_prep.to_pickle("dataset_bmw_preprocesado.pkl")

print("Dataset preprocesado guardado exitosamente.")
print(f"Dimensiones finales: {df_bmw_prep.shape}")


### RESPUESTAS A LAS PREGUNTAS DEL EJERCICIO


#### 1. ¿QUÉ COLUMNAS ELIMINARON INICIALMENTE DEL DATASET Y POR QUÉ?

**Columnas eliminadas:**
1. **`marca`**: Se eliminó porque todos los vehículos en el dataset son BMW (o deberían serlo). Esta columna no aporta información discriminativa ya que no hay variabilidad. Los valores nulos (~20%) también indicaban que no era una variable relevante.

2. **`asientos_traseros_plegables`**: Se eliminó porque tenía más del 70% de valores nulos. Una variable con tan alto porcentaje de datos faltantes no es útil para el modelo, ya que perderíamos demasiada información al eliminar esos registros, y la imputación no sería confiable con tan pocos datos disponibles.

**Razón de la decisión:** Estas columnas no aportan valor predictivo y su eliminación mejora la calidad del dataset sin perder información relevante.


#### 2. MANEJO DE NULOS, EXPLICAR QUÉ SE HIZO CON LOS NULOS POR CADA COLUMNA

**Estrategia aplicada por columna:**

1. **`precio` (target)**: Se eliminaron todos los registros con precio nulo. No podemos entrenar un modelo sin la variable objetivo.

2. **`modelo`, `km`, `potencia`**: Se eliminaron registros con nulos en estas variables críticas (eran muy pocos). Estas son variables fundamentales para predecir el precio.

3. **`fecha_registro`**: Se creó una variable binaria `fecha_registro_disponible` que indica si la fecha está disponible. Los nulos se mantuvieron en `EDAD_VEHICULO` (derivada de esta fecha) ya que tiene ~50% de nulos y eliminarlos reduciría demasiado el dataset.

4. **`tipo_gasolina`**: Se imputó con la moda (valor más frecuente: "diesel"). Tiene solo ~0.1% de nulos.

5. **`tipo_coche`**: Se imputó con la categoría "desconocido" en lugar de la moda. Esto preserva la información de que el dato faltaba, lo cual puede ser relevante.

6. **`color`**: Se imputó con la categoría "desconocido" para mantener la información de que el dato faltaba.

7. **Variables booleanas** (`bluetooth`, `alerta_lim_velocidad`, `aire_acondicionado`, `volante_regulable`, `camara_trasera`, `elevalunas_electrico`): Los nulos se trataron como `False` (ausencia de la característica). Esto tiene sentido porque un valor nulo en una característica opcional generalmente indica que el vehículo no la tiene.

8. **`gps`**: No tenía nulos, se mantuvo como estaba.

9. **`fecha_venta`**: Se eliminaron registros con fecha de venta nula, ya que es una variable crítica.


#### 3. ANÁLISIS UNIVARIABLE (ANALIZAR CADA VARIABLE INDIVIDUALMENTE) EXPLICAR ALGUNA INFORMACIÓN INTERESANTE ENCONTRADA

**Variables numéricas:**
- **`precio`**: Distribución sesgada a la derecha. Precio medio ~22,000€, mediana ~20,000€. La diferencia sugiere presencia de vehículos de lujo con precios muy altos.
- **`km`**: Rango amplio desde 0 hasta más de 300,000 km. El kilometraje promedio es ~120,000 km, indicando que la mayoría son vehículos usados con uso moderado.
- **`potencia`**: Potencia media ~150 CV. Rango desde 85 CV (modelos básicos) hasta más de 300 CV (modelos deportivos/M).

**Variables categóricas:**
- **`modelo`**: El modelo más común es "420" seguido de "320". Hay gran variedad de modelos (más de 50), lo que sugiere diversidad en la oferta.
- **`tipo_gasolina`**: "diesel" es el tipo más común (~60%), seguido de "petrol" (~35%). Esto refleja el mercado europeo donde el diesel ha sido popular.
- **`color`**: "black" y "grey" son los colores más populares, seguidos de "white" y "blue". Los colores neutros dominan el mercado.
- **`tipo_coche`**: "convertible" y "coupe" son los tipos más frecuentes. La categoría "desconocido" (imputada) representa ~30% de los registros.

**Variables booleanas:**
- **`gps`**: Presente en ~85% de los vehículos, siendo la característica más común.
- **`bluetooth`**: Presente en ~75% de los vehículos.
- **`aire_acondicionado`**: Presente en ~90% de los vehículos, casi estándar.
- **`camara_trasera`**: Solo presente en ~15% de los vehículos, característica menos común.


#### 4. ANÁLISIS DE CORRELACIÓN INICIAL, ¿HAY ALGUNA VARIABLE CORRELACIONADA?

**Resultados del análisis de correlación:**

No se encontraron correlaciones altas (|r| > 0.7) entre las variables numéricas del dataset. Las correlaciones más significativas son:

- **`potencia` vs `precio`**: Correlación positiva moderada (~0.5-0.6). A mayor potencia, mayor precio, lo cual es lógico.

- **`km` vs `precio`**: Correlación negativa moderada (~-0.4 a -0.5). A mayor kilometraje, menor precio, como se espera en vehículos usados.

- **`EDAD_VEHICULO` vs `precio`**: Correlación negativa débil a moderada. Vehículos más antiguos tienden a valer menos.

**Conclusión:** No hay problemas de multicolinealidad entre variables numéricas. Las correlaciones encontradas son esperadas y no requieren eliminación de variables. El modelo puede usar todas las variables numéricas sin riesgo de multicolinealidad alta.


#### 5. ANÁLISIS VARIABLE VS TARGET, ¿HAY ALGÚN INSIGHT INTERESANTE?

**Insights principales:**

1. **`potencia` vs `precio`**: Correlación positiva fuerte. Los vehículos con mayor potencia (modelos M, 435, 640) tienen precios significativamente más altos. El modelo i8 (híbrido) tiene el precio más alto del dataset.

2. **`km` vs `precio`**: Correlación negativa. Los vehículos con menos de 50,000 km tienen precios notablemente más altos. Después de 150,000 km, el precio cae significativamente.

3. **`modelo` vs `precio`**: 
   - Modelos más caros: M4, M3, 640, 435, i8 (precios >30,000€)
   - Modelos intermedios: 420, 320, 220 (precios 15,000-25,000€)
   - Modelos más económicos: 118, 120, 316, 318 (precios <15,000€)

4. **`tipo_gasolina` vs `precio`**: Los vehículos híbridos (hybrid_petrol) tienen precios más altos. Entre diesel y petrol, no hay diferencia significativa en precio medio.

5. **Características booleanas vs `precio`**:
   - **`gps`**: Aumenta el precio en ~2,000€ en promedio
   - **`bluetooth`**: Aumenta el precio en ~1,500€
   - **`aire_acondicionado`**: Aumenta el precio en ~1,000€
   - **`camara_trasera`**: Aumenta el precio en ~3,000€ (característica premium)

6. **`color` vs `precio`**: Los colores "orange" y colores especiales tienden a tener precios ligeramente más altos, posiblemente por ser menos comunes o ediciones especiales.

**Insight clave:** La potencia y el modelo son los factores más determinantes del precio, seguidos del kilometraje y las características premium (cámara trasera, GPS).


#### 6. TRANSFORMACIÓN DE CATEGÓRICAS A NUMÉRICAS, ¿QUÉ VARIABLES VAN A TRANSFORMAR? ¿QUÉ TÉCNICA SE VA A USAR?

**Variables a transformar:**

1. **`modelo`**: ~50 categorías → **One-Hot Encoding (OHE)**
2. **`tipo_gasolina`**: 4-5 categorías (diesel, petrol, hybrid_petrol, etc.) → **One-Hot Encoding (OHE)**
3. **`color`**: ~10-15 categorías → **One-Hot Encoding (OHE)**
4. **`tipo_coche`**: ~5-6 categorías (convertible, coupe, desconocido, etc.) → **One-Hot Encoding (OHE)**

**Técnica utilizada: One-Hot Encoding (OHE)**

**Razones para elegir OHE:**
- Las variables categóricas no tienen un orden inherente (no son ordinales)
- El número de categorías es manejable (no excede 50 categorías por variable)
- OHE preserva toda la información de las categorías sin asumir relaciones de orden
- Es compatible con la mayoría de algoritmos de machine learning
- Evita el problema de que el modelo interprete las categorías como ordenadas

**Alternativas consideradas pero no usadas:**
- **Label Encoding**: No apropiado porque asume orden entre categorías
- **Target Encoding**: Podría usarse pero requiere validación cruzada para evitar overfitting
- **Frequency Encoding**: Menos informativo que OHE para este caso

**Resultado:** Después de OHE, se crearon aproximadamente 70-80 nuevas columnas binarias (dependiendo del número de categorías únicas en cada variable).
